compiler/rustc_builtin_macros/src/deriving/generic/mod.rs

```diff
@@ -475,6 +475,26 @@ impl<'a> TraitDef<'a> {
                     false
                 });
 
+                if let ast::ItemKind::Struct(struct_def, _) | ast::ItemKind::Union(struct_def, _) =
+                    &item.kind
+                    && let Some(&trait_name) = self.path.components().last()
+                    && !(trait_name == sym::Copy || trait_name == sym::Clone)
```

Comment:
Theoretically we could move the “`#[derive(Clone)]` requires accompanying `#[derive(Copy)]`” check here to avoid the duplication, right? Or you do think this “violates the layers of abstraction”? I haven't thought that hard about it.

Ah, but that would lead to us performing more work in the error path since we wouldn't be able to return early in `expand_deriving_clone`. However, that should be fine as long as we don't trigger any debug asserts.

Reply:
Ah, one more thing: Could you add a `FIXME(unnamed_fields)` (here or in mod `clone` depending on the way we go forward) saying that we should permit `derive(Clone)` without `derive(Copy)` once we can support it, i.e., once struct initialization supports unnamed fields.